tutorials/getting_started/laziness.md

```diff
@@ -0,0 +1,233 @@
+        self._token_indexers = token_indexers or {'tokens': SingleIdTokenIndexer()}
+```
+
+And its `from_params` looks for a `lazy` configuration parameter
+(defaulting to `False` if it's not specified):
+
+```python
+    @classmethod
+    def from_params(cls, params: Params) -> 'SnliReader':
+        tokenizer = Tokenizer.from_params(params.pop('tokenizer', {}))
+        token_indexers = TokenIndexer.dict_from_params(params.pop('token_indexers', {}))
+        lazy = params.pop('lazy', False)
+        params.assert_empty(cls.__name__)
+        return SnliReader(tokenizer=tokenizer,
+                          token_indexers=token_indexers,
+                          lazy=lazy)
+```
+
+Any dataset reader that you want to handle lazy datasets should behave likewise.
+
+## Laziness in `DatasetReader.read`
+
+The primary public interface of the `DatasetReader` is its `read` method, which is implemented in
+the base class (and which I've stripped down to its essence):
+
+```python
+    def read(self, file_path: str) -> Iterable[Instance]:
+        if self.lazy:
+            return _LazyInstances(lambda: iter(self._read(file_path)))
+        else:
+            return ensure_list(self._read(file_path))
+```
+
+In both cases, it calls the private `_read` method, which you have to implement,
+and which itself returns an `Iterable[Instance]`. More on that below.
+
+If the dataset reader was instantiated with `lazy=False`,
+it just makes sure that the returned instances are in a `list` and returns that list.
+That list is loaded into memory and (because it's a list) can be iterated over repeatedly
+after the initial call to `.read()`.
+
+The more interesting case is when the dataset reader was instantiated with `lazy=True`.
+In that case we return a `_LazyInstances` initialized with a lambda function
+that calls `_read()`. `_LazyInstances` is just a simple class wrapper to produce
+lazy iterables (again, I stripped it down to just its essence):
+
+```python
+class _LazyInstances(Iterable):
+    def __init__(self, instance_generator: Callable[[], Iterator[Instance]]) -> None:
+        self.instance_generator = instance_generator
+
+    def __iter__(self) -> Iterator[Instance]:
+        return self.instance_generator()
+```
+
+What this means is that the result is an iterable that calls the provided
+function each time it's iterated over. With the lambda we passed it,
+it will call `self._read(file_path)` each time it's iterated over.
+
+In other words, if you implement your own dataset reader and do something like the following:
+
+```python
+reader = MyDatasetReader(lazy=True)
+instances = reader.read('my_instances.txt')
+for epoch in range(10):
+    for instance in instances:
+        process(instance)
+```
+
+Then each epoch's `for instance in instances` results in a *new* call
+to `MyDatasetReader._read()`, and your instances will be read from disk
+10 times.
+
+In this case, you should make sure that your `_read()` method
```

Comment:
I'm not sure what the right way is to talk about this, but if you're sampling things at each call to `_read()`, you actually _do_ want to read the whole dataset into memory, but only on the _first_ call to `_read()`.  Then, after reading everything that you want, all calls to `_read()` do the sampling on the instances in memory, and it's different at each epoch.  This is an advanced usage that you might not want to cover here, but I'd at least avoid contradicting this use case in the simple tutorial.

Reply:
(1) why would you implement sampling in the DatasetReader rather than as a `SamplingIterator` or something?
(2) doesn't this depend on how you sample? you could take each instance with a fixed probability without loading the whole dataset, or you could do some kind of reservoir sampling.
(3) all that said, I guess I can imagine cases where you would want to sample in the DatasetReader, so let me think about how to word that

Reply:
The case I'm thinking of is that you have `k`paragraphs associated with each instance, and at each epoch you want to sample 2 of them. You read in all `k`, cache that, then pick two at each call to`_read`. You can't easily do this kind of thing in our `DataIterator` - you're already dealing with`Instances`by that point.

And by the way, this is how I recommend doing a `DatasetReader` for Chris's model.